File: discord/src/main/java/io/github/shaksternano/borgar/discord/media/io/reader/LimitedDurationMediaReader.java
package io.github.shaksternano.borgar.discord.media.io.reader;

import io.github.shaksternano.borgar.discord.media.VideoFrame;
import io.github.shaksternano.borgar.discord.util.collect.ClosableIterator;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.NoSuchElementException;

public class LimitedDurationMediaReader<E extends VideoFrame<?, E>> extends BaseMediaReader<E> {

    private final MediaReader<E> reader;
    @Nullable
    private LimitedDurationMediaReader<E> reversed;

    public LimitedDurationMediaReader(MediaReader<E> reader, long maxDuration) throws IOException {
        super(reader.format());
        this.reader = reader;
        var readerInfo = readerInfo(reader, maxDuration);
        frameCount = readerInfo.frameCount();
        this.duration = readerInfo.duration();
        frameDuration = (double) duration / frameCount;
        frameRate = 1_000_000 / frameDuration;
        width = reader.width();
        height = reader.height();
        loopCount = reader.loopCount();
    }

    private ReaderInfo readerInfo(MediaReader<E> reader, long maxDuration) throws IOException {
        try (var iterator = reader.iterator()) {
            var frameCount = 0;
            var duration = 0L;
            while (iterator.hasNext()) {
                var frame = iterator.next();
                var newDuration = duration + (long) frame.duration();
                if (newDuration > maxDuration) {
                    if (frameCount == 0) {
                        return new ReaderInfo(1, newDuration);
                    } else {
                        return new ReaderInfo(frameCount, duration);
                    }
                }
                frameCount++;
                duration = newDuration;
            }
        }
        return new ReaderInfo(reader);
    }

    @Override
    public E readFrame(long timestamp) throws IOException {
        var circularTimestamp = timestamp % Math.max(duration(), 1);
        return reader.readFrame(circularTimestamp);
    }

    @Override
    public MediaReader<E> reversed() throws IOException {
        if (reversed == null) {
            reversed = new LimitedDurationMediaReader<>(reader.reversed(), duration);
            reversed.reversed = this;
        }
        return reversed;
    }

    @Override
    public ClosableIterator<E> iterator() {
        return new Iterator();
    }

    @Override
    public void close() throws IOException {
        reader.close();
    }

    private class Iterator implements ClosableIterator<E> {

        private final ClosableIterator<E> delegate = reader.iterator();
        private long currentTimestamp = 0;

        @Override
        public boolean hasNext() {
            return currentTimestamp < duration && delegate.hasNext();
        }

        @Override
        public E next() {
            if (currentTimestamp >= duration) {
                throw new NoSuchElementException("Iterator has no more elements");
            }
            var frame = delegate.next();
            currentTimestamp += frame.duration();
            return frame;
        }

        @Override
        public void close() throws IOException {
            delegate.close();
        }
    }

    private record ReaderInfo(int frameCount, long duration) {

        private ReaderInfo(MediaReader<?> reader) {
            this(reader.frameCount(), reader.duration());
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/reader/LimitedDurationMediaReader.kt
package io.github.shaksternano.borgar.core.media.reader

import io.github.shaksternano.borgar.core.media.VideoFrame
import io.github.shaksternano.borgar.core.util.circular
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.takeWhile
import kotlin.time.Duration

class LimitedDurationMediaReader<T : VideoFrame<*>> internal constructor(
    private val reader: MediaReader<T>,
    readerInfo: ReaderInfo,
) : BaseMediaReader<T>() {

    override val frameCount: Int = readerInfo.frameCount
    override val duration: Duration = readerInfo.duration
    override val frameDuration: Duration = duration / frameCount
    override val frameRate: Double = 1000.0 / frameDuration.inWholeMilliseconds
    override val audioChannels: Int = reader.audioChannels
    override val audioSampleRate: Int = reader.audioSampleRate
    override val audioBitrate: Int = reader.audioBitrate
    override val width: Int = reader.width
    override val height: Int = reader.height
    override val loopCount: Int = reader.loopCount

    override suspend fun readFrame(timestamp: Duration): T {
        val circularTimestamp = timestamp.circular(duration)
        return reader.readFrame(circularTimestamp)
    }

    override fun asFlow(): Flow<T> {
        var currentTimestamp = Duration.ZERO
        return reader.asFlow().takeWhile {
            if (currentTimestamp >= duration) return@takeWhile false
            currentTimestamp += it.duration
            true
        }
    }

    override fun createReversed(): MediaReader<T> =
        LimitedDurationMediaReader(
            reader.reversed,
            ReaderInfo(reader),
        )

    override suspend fun close() = reader.close()
}

@Suppress("FunctionName")
suspend fun <T : VideoFrame<*>> LimitedDurationMediaReader(
    reader: MediaReader<T>,
    duration: Duration,
): MediaReader<T> = LimitedDurationMediaReader(
    reader,
    readerInfo(reader, duration),
)

private suspend fun readerInfo(reader: MediaReader<*>, maxDuration: Duration): ReaderInfo {
    if (reader.isEmpty) return ReaderInfo(reader)
    var frameCount = 0
    var totalDuration = Duration.ZERO
    reader.asFlow().takeWhile {
        val newDuration = totalDuration + it.duration
        if (newDuration > maxDuration) {
            if (frameCount == 0) {
                frameCount = 1
                totalDuration = newDuration
            }
            return@takeWhile false
        }
        frameCount++
        totalDuration = newDuration
        return@takeWhile true
    }.collect()
    return ReaderInfo(frameCount, totalDuration)
}

internal data class ReaderInfo(val frameCount: Int, val duration: Duration) {
    constructor(reader: MediaReader<*>) : this(
        reader.frameCount,
        reader.duration,
    )
}
